New speed chooser ui

lks-nbg opened the pull request:
This PR is a complete redesign of the current speed chooser. Instead of selecting options from a given list the user can now choose a custom playback speed between 0.5 and 3.0 in 0.05 steps with a slider.

![Screenshot_20220330-175956_2](https://user-images.githubusercontent.com/42645195/160882178-24e43a2e-b24a-4902-a365-afed309ec96d.jpg)
 
During testing I received some leak notifications, but I am not sure if this was caused by this PR. It should definitely be thoroughly checked and tested. Furthermore I noticed that sometimes the slider value is stuck and does not get updated. The actual playback speed and the button in the player however are working as expected. 

One thing to mention is that the current implementation relies on materialLibVersion = '1.6.0-beta01' which allows to show the slider label all the time and not only when touched (app:labelBehavior="visible"). If this dependency is not accepted, we can change that and start with the progress only visible while sliding and change it as soon as 1.6.0 is published as stable.

Resolves #7 / Alternative to #63

Commit:
New speed chooser ui

Commit:
Fix for BottomNavigationView label above icon

mattttvaughn commented:
The new leak notifications have been around for a few days fyi, related to a toast somewhere if i recall but haven't looked into it yet

mattttvaughn commented on app/src/main/java/io/github/mattpvaughn/chronicle/features/currentlyplaying/CurrentlyPlayingViewModel.kt:
```diff
@@ -260,7 +271,10 @@ class CurrentlyPlayingViewModel(
         when (key) {
-            PrefsRepo.KEY_PLAYBACK_SPEED -> _speed.postValue(prefsRepo.playbackSpeed)
+            PrefsRepo.KEY_PLAYBACK_SPEED -> _speed.postValue(
+                if(prefsRepo.playbackSpeed < PLAYBACK_SPEED_MIN || prefsRepo.playbackSpeed > PLAYBACK_SPEED_MAX) PLAYBACK_SPEED_DEFAULT
```
This is very edge-case-y, but we could possibly run into some concurrency issues here since this technically pulls from SharedPreferences on each `prefsRepo.playbackSpeed`.

Can just do a prefsRepo.playbackSpeed once at the top and save to a variable for safety

mattttvaughn commented on app/src/main/java/io/github/mattpvaughn/chronicle/views/ModalBottomSheetSpeedChooser.kt:
```diff
@@ -0,0 +1,80 @@
+                "1.2x" -> 1.2f
+                "1.5x" -> 1.5f
+                "2.0x" -> 2.0f
+                // R.string.playback_speed_1_0x -> 1.0f  (not sure why this gives an error)
```
wow yea that's weird. looks like Android is doing something whacky under the hood. I'm okay with hardcoding if you don't find a better way to do it

mattttvaughn commented on app/src/main/java/io/github/mattpvaughn/chronicle/features/currentlyplaying/CurrentlyPlayingViewModel.kt:
```diff
@@ -616,42 +630,19 @@ class CurrentlyPlayingViewModel(
-                }
-            }
-        )
+        modalBottomSheetSpeedChooser.show(fragmentManager, ModalBottomSheetSpeedChooser.TAG)
```
The fragmentManager is scoped to the activity, while the ViewModel persists across activity changes, so we get into memory leak and destroyed view trouble by doing this. (you can force a crash in this case by opening the speed chooser, rotating device, then speed choose again)

To get around this we can expose the bottom chooser "shown" state (could be as simple as a LiveData<Boolean> + MutableLiveData<Boolean>) from the viewmodel. Then in fragment or databinding we observe that data and use it to hide/show the speed chooser

mattttvaughn commented on build.gradle:
```diff
@@ -7,7 +7,7 @@ buildscript {
         gradleVersion = '7.0.2'
         supportlibVersion = '1.2.0'
-        materialLibVersion = '1.3.0'
+        materialLibVersion = '1.6.0-beta01' // todo: permanent visible slider label is available from 1.6 (currently in beta) → https://github.com/material-components/material-components-android/releases
```
appreciate the callout on why this got bumped 👍

LGTM

mattttvaughn commented on app/src/main/java/io/github/mattpvaughn/chronicle/features/currentlyplaying/CurrentlyPlayingViewModel.kt:
```diff
@@ -246,6 +255,8 @@ class CurrentlyPlayingViewModel(
     val bottomChooserState: LiveData<BottomChooserState>
         get() = _bottomChooserState
 
+    private val modalBottomSheetSpeedChooser = ModalBottomSheetSpeedChooser()
```
`View`s should live in the fragment since they should are bound to fragment/activity lifecycles. also helps with testability

more details in my comment below about the FragmentManager

mattttvaughn commented on app/src/main/res/layout/modal_bottom_sheet_speed_chooser.xml:
```diff
@@ -0,0 +1,146 @@
+<?xml version="1.0" encoding="utf-8"?>
+<layout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto">
```
looks amazing!

Commit:
New speed chooser ui

lks-nbg commented:
Thanks for reviewing. I made the suggested changes. 

But I am still having trouble updating the change_speed_button. I added observers for speed & playbackSpeedString for debugging and noticed that these variables do not change after updatePlaybackSpeed is called. Any ideas on how to solve this?

mattttvaughn committed:
Minor speed chooser fixes

 - Fix memory leak
 - Fix multiple modals being shown over each other
 - Fix occasional crash caused by multiple CurrentlyPlayingViewModels getting instantiated

mattttvaughn commented:
Pushed a commit which should resolve a few issues. Apologies for not letting you working things out but I ended up making most of the changes while debugging the issue and figured better to not let it go to waste.

Will add some comments to the PR explaining the issues they resolve

Feel free to review it if you see any issues

mattttvaughn commented on app/src/main/java/io/github/mattpvaughn/chronicle/features/currentlyplaying/CurrentlyPlayingFragment.kt:
```diff
@@ -125,17 +123,14 @@ class CurrentlyPlayingFragment : Fragment() {
-            Timber.i("viewModel.playbackSpeedString.observe → $value")
-            // todo: remove → only for debugging
+        viewModel.showModalBottomSheetSpeedChooser.observe(viewLifecycleOwner) { eventShowChooser ->
+            if (!eventShowChooser.hasBeenHandled) {
```
emitting this as an event since the `ModalBottomSheetSpeedChooser` is a fragment and survives config changes. event is cleared by `getContentIfNotHandled`

Minor issue, but previously, if you rotated the screen over and over (causing `CurrentlyPlayingFragment` to get recreated over and over and keep showing a new `ModalBottomSheetChooser` over and over), a bunch of choosers would stack up on top of each other.

mattttvaughn commented on app/src/main/java/io/github/mattpvaughn/chronicle/views/ModalBottomSheetSpeedChooser.kt:
```diff
@@ -1,31 +1,29 @@
 class ModalBottomSheetSpeedChooser : BottomSheetDialogFragment() {
 
-    @Inject
-    lateinit var viewModelFactory: CurrentlyPlayingViewModel.Factory
```
I'm pretty sure since `CurrentlyPlayingViewModel.Factory` isn't specifically scoped, the viewModel we injecting here was a different instance than the one in `CurrentlyPlayingViewModelFragment`.

Ended up pulling this out anyways since it's probably better for the two to not be coupled anyways. Now we can add a `ModalBottomSheetSpeedChooser` anywhere in the app without needing a `CurrentlyPlayingViewModel`

mattttvaughn commented on app/src/main/java/io/github/mattpvaughn/chronicle/features/currentlyplaying/CurrentlyPlayingViewModel.kt:
```diff
@@ -127,16 +130,21 @@ class CurrentlyPlayingViewModel(
-    private var _speed = MutableLiveData(prefsRepo.playbackSpeed)
-    val speed: LiveData<Float>
-        get() = _speed
+    val speed = FloatPreferenceLiveData(
```
just a minor stylistic choice tbh

lks-nbg commented:
> Pushed a commit which should resolve a few issues. Apologies for not letting you working things out but I ended up making most of the changes while debugging the issue and figured better to not let it go to waste.

No worries. I reached my limits and was not able to finish it myself. So I am grateful and happy to learn from your experience. The code looks good to me and during testing everything worked as expected 👍 
From my side, this can be merged to get user feedback.
